### PR TITLE
[clang] Emit nuw GEPs for array subscript expressions

### PR DESCRIPTION
**Pull request by hazzlim**
Generate nuw GEPs for array subscript expressions where the base address points to the base of a constant size array and the index is unsigned.

### CODE REVIEW

**Comment by hazzlim** [comment]
This seems correct from my reading of the relevant parts of the C/C++ standard, regarding pointer arithmetic and array subscript expressions.

It does actually seem like the restriction to unsigned indices here is actually unnecessary, and adding `nuw` is also valid for signed indices here - but I wanted to check if people agree on that.

**Comment by efriedma-quic** [comment]
> adding nuw is also valid for signed indices here

I don't understand how you think this would work; a-1 and a+-1 are required to produce the same result.

File: clang/lib/CodeGen/CGExpr.cpp [hunk 1]
```diff
@@ -4345,8 +4355,8 @@ LValue CodeGenFunction::EmitArraySubscriptExpr(const ArraySubscriptExpr *E,
     QualType ptrType = E->getBase()->getType();
     Addr = emitArraySubscriptGEP(*this, Addr, Idx, E->getType(),
                                  !getLangOpts().isSignedOverflowDefined(),
-                                 SignedIndices, E->getExprLoc(), &ptrType,
-                                 E->getBase());
+                                 /*nuw=*/false, SignedIndices, E->getExprLoc(),
+                                 &ptrType, E->getBase());
```
Comment (nikic) [inline clang/lib/CodeGen/CGExpr.cpp:4359]: Why is the handling for arrays and pointers here different? Aren't they essentially semantically equivalent?

**Review by nikic — COMMENTED** [review]
As a high level comment: Do we need to pass through the nuw flag, rather than deriving it from signedIndices?

Basically, I think that we should be emitting nuw if and only if `-fsanitize=undefined` would perform a check for unsigned overflow for that gep, and the information to determine whether that is done should already be available in the relevant places.

(Unless we currently perform incorrect sanitizer checks in some cases -- in which case that needs to be fixed.)

**Comment by hazzlim** [comment]
> As a high level comment: Do we need to pass through the nuw flag, rather than deriving it from signedIndices?
> 
> Basically, I think that we should be emitting nuw if and only if `-fsanitize=undefined` would perform a check for unsigned overflow for that gep, and the information to determine whether that is done should already be available in the relevant places.
> 
> (Unless we currently perform incorrect sanitizer checks in some cases -- in which case that needs to be fixed.)

Abandoning this PR in favor of taking this approach in #105496

**Comment by hazzlim** [comment]
> > adding nuw is also valid for signed indices here
> 
> I don't understand how you think this would work; a-1 and a+-1 are required to produce the same result.

The thinking here RE signed indices was that in this special case, where the base address of the GEP is the start of an array, the index cannot be negative without violating the `inbounds` attribute (alternatively the requirement of a pointer arithmetic expression to evaluate to an element of the array or one past the end, in C/C++ standard terms). Therefore we could prove `nuw` via `inbounds` + `nneg`.

I wonder if others agree that this is valid or can see an issue with this?

**Comment by efriedma-quic** [comment]
Oh, I see.

If you can prove that it would be illegal to use a negative index with a particular pointer, you might be able to optimize... but I'm not sure how aggressive we want to be here.  Even if the C standard technically disallows indexing past the beginning/end of a member array, it's not something we've historically checked or optimized, so it's likely to break code.

**Comment by nikic** [comment]
Can this be closed in favor of https://github.com/llvm/llvm-project/pull/105496?

**Comment by hazzlim** [comment]
> Can this be closed in favor of #105496?

Yes good point - will close this and open a PR to Reland #105496.

**Pull request closed without merging**